### PR TITLE
Introduce color picker to context pad

### PR DESCRIPTION
**Pull request by smbea**
Related to #1491
Design issue: https://github.com/camunda/camunda-modeler-design/issues/101

Screenshot:
Screenshot 2022-05-19 at 13 53 06

### CODE REVIEW

File: lib/features/popup-menu/ColorMenuProvider.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. Licensed under a commercial license.
+ * You may not use this file except in compliance with the commercial license.
+ */
+
+var COLORS = [{
```
Comment (nikku) [inline lib/features/popup-menu/ColorMenuProvider.js:7]: We should align closely here with [bpmn-js-color-picker](https://github.com/bpmn-io/bpmn-js-color-picker/blob/master/colors/ColorPopupProvider.js#L20), including but not limited to 

* Do not copy Camunda modeler code (proprietary license)
* Clear colors when the user sets "white"
* Potential other topics (i'm not aware of any to date)

Comment (smbea) [reply lib/features/popup-menu/ColorMenuProvider.js:1]: > Do not copy Camunda modeler code (proprietary license header)

Oops, I fixed thats.

> Clear colors when the user sets "white"

Also addressed this.

I also took a look at the bpmn-js-color-picker, and didn't notice any other discrepancies

Comment (nikku) [reply lib/features/popup-menu/ColorMenuProvider.js:1]: :+1:

**Comment by philippfromme** [comment]
Have we settled on a UI design already? I feel like it could be improved.

**Comment by smbea** [comment]
> Have we settled on a UI design already? I feel like it could be improved.

This design mimics what was done on the web-modeler, which was linked in the design issue. But I guess it's not set in stone.

**Comment by philippfromme** [comment]
I was just wondering since the color picker of the web modeler was created a long time ago and I believe it would look a bit different was it created today. Were there any discussion about whether to keep rolling with that design or not?

**Comment by smbea** [comment]
> I was just wondering since the color picker of the web modeler was created a long time ago and I believe it would look a bit different was it created today. Were there any discussion about whether to keep rolling with that design or not?

There wasn't, as far as I'm concerned. We can keep this open and have @andreasgeier give his opinion when he's back

**Comment by philippfromme** [comment]
I'm not trying to block this. I'm fine with going forward now, but I guess giving it a little overhaul at some point would be nice. 🙂

**Comment by smbea** [comment]
@philippfromme Let's move on with this now then. I will leave a comment on the design issue so it's not forgotten, and we can re-evaluate the design at a later stage.

File: assets/bpmn-js.css [hunk 2]
```diff
@@ -113,4 +113,61 @@
 
 .selected .bjs-drilldown-empty {
   display: inherit;
+}
+
+/* COLOR PICKER */
+
+.bpmn-color-picker .djs-popup-header {
```
Comment (nikku) [inline assets/bpmn-js.css:120]: We should namespace _all_ classes (`bjs-***`).

File: lib/features/popup-menu/ColorMenuProvider.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+ColorMenuProvider.prototype.getEntries = function() {
+  return [];
+};
+
+ColorMenuProvider.prototype.getHeaderEntries = function(elements) {
+  var modeling = this._modeling;
+  var translate = this._translate;
+
+  return COLORS.map(function(color) {
```
Comment (pinussilvestrus) [inline lib/features/popup-menu/ColorMenuProvider.js:48]: Do we consider making the colors configurable via `config.colors` or something?

Comment (smbea) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: I hadn't considered. We would need to make sure the color menu UI would support this. The menu colors are set in the CSS file at the moment, which wouldn't be possible anymore if the colors aren't fixed. And we would need to handle a variable amount of colours.

Comment (nikku) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: I think it is fine to not make these variable at the moment, but also not 100% block the door regarding this.

What would be most important for me is that we have a story around "how customization is done". A number of options are on the table:

* We add the color picker per default, but make it extensible (based on what exactly?)
* We improve the existing bpmn-js-color-picker (with multi-element support); establish that as our tool for the job (and make that extensible)

Either was the result is the same: Our users can resort to a way to change colors on the diagram, both supporting single element coloring, but also grouping (new with multi-element context pad!).

I'd be absolutely fine with simply improving the existing color picker for the moment.

Comment (barmac) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: I think it's OK to have it as is. Configurability can be a separate feature -> implemented via another issue.

Comment (nikku) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: Generally speaking we shall not support arbitary colors; this is how our tooling works; this is how we ensure users use colors to transport semantics (their choice) but not shoot them into the foot.

File: assets/bpmn-js.css (hunk 2 above)
Comment (smbea) [reply assets/bpmn-js.css:120]: `bpmn-color-picker` is derived from `popupMenu.registerProvider('bpmn-color-picker', _)`. I name it like this following the convention of `bpmn-replace`

Comment (nikku) [reply assets/bpmn-js.css:120]: You could call it `bjs-color-picker` or whatever, but we should always namespace new additions to our CSS.

File: lib/features/popup-menu/ColorMenuProvider.js (hunk 3 above)
Comment (nikku) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: > I think it's OK to have it as is. Configurability can be a separate feature -> implemented via another issue.

But if we can go for a simple, no CSS solution to do the job, that would be great, too.

Maybe we can generate these icons as an SVG on the fly? This way we'd be able to support arbitary colors (some day maybe) but also not require CSS.

Comment (barmac) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: True. I haven't noticed that the color boxes are partially defined in the stylesheet. Ideally, I'd expect the fill and stroke values to be defined in hash style due to how it is defined by the bpmn-in-color extension (`#xxxxxx`), and these values should be used to draw the icon.

Comment (barmac) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: Without changes in diagram-js, we could use the `imageUrl` property on the popup entry to pass a data URL containing the colored icon: https://github.com/bpmn-io/diagram-js/blob/20ab8b869bea9dc8f7b6e9a1f98daa0eb1fe358c/lib/features/popup-menu/PopupMenu.js#L540

Comment (nikku) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: Exactly exactly. :slightly_smiling_face:

Comment (smbea) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: Maybe we should a proper discussion on this. Otherwise this discussion could get very long here

Comment (nikku) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: Let me summarize the feedback you got:

* We should not rely on CSS for "coloring of entries", but generate images (via data-URL) and a (configurable) set of colors.
* This solves two issues: Colors should be defined in a single place; colors should be easily configurable (some day maybe).

----

Is there anything else we want to discuss? If so, let us do it.

Comment (nikku) [reply lib/features/popup-menu/ColorMenuProvider.js:48]: Proposal for configuration:

```
// disable color picker
new BpmnJS({
  colorPicker: false
});

// customize colors
new BpmnJS({
  colorPicker: {
    colors: [ { label, stroke, fill }* ]
});
```

**Comment by smbea** [comment]
**Summary of discussion:**
* Keep color-picker out of core bpmn-js. Make use of existing `bpmn-js-color-picker` extension and include it in `camunda-bpmn-js`. This way, users can add it only if they're interested in using it (cc @christian-konrad)
* Generate color icons (via data-URL) based on defined colors -> instead of defining it with CSS
* (Now or create an issue for later) allow `colors` to be configurable

So I will close this PR . I created an issue to move this progress to `bpmn-js-color-picker` - https://github.com/bpmn-io/bpmn-js-color-picker/issues/8

**Comment by christian-konrad** [comment]
> * Keep color-picker out of core bpmn-js. Make use of existing `bpmn-js-color-picker` extension and include it in `camunda-bpmn-js`. This way, users can add it only if they're interested in using it (cc @christian-konrad)

Great, let's keep the core as lean as possible.

**Pull request closed without merging**